src/zeroD/IdealGasMoleReactor.cpp

```diff
@@ -0,0 +1,237 @@
+        // production in gas phase and from surfaces
+        dydt[n] = (m_wdot[n] * m_vol + m_sdot[n]);
+    }
+
+    // add terms for outlets
+    for (auto outlet : m_outlet) {
+        for (size_t n = 0; n < m_nsp; n++) {
+            // flow of species into system and dilution by other species
+            dydt[n] -= outlet->outletSpeciesMassFlowRate(n) * imw[n];
+        }
+        double mdot = outlet->massFlowRate();
+        mcvdTdt -= mdot * m_pressure * m_vol / m_mass; // flow work
+    }
+
+    // add terms for inlets
+    for (auto inlet : m_inlet) {
+        double mdot = inlet->massFlowRate();
+        mcvdTdt += inlet->enthalpy_mass() * mdot;
+        for (size_t n = 0; n < m_nsp; n++) {
+            double mdot_spec = inlet->outletSpeciesMassFlowRate(n);
+            // flow of species into system and dilution by other species
+            dydt[n] += inlet->outletSpeciesMassFlowRate(n) * imw[n];
+            mcvdTdt -= m_uk[n] * imw[n] * mdot_spec;
+        }
+    }
+
+    RHS[1] = m_vdot;
+    if (m_energy) {
+        RHS[0] = mcvdTdt / (m_mass * m_thermo->cv_mass());
+    } else {
+        RHS[0] = 0.0;
+    }
+}
+
+Eigen::SparseMatrix<double> IdealGasMoleReactor::jacobian(double t, double* LHS, double* RHS)
+{
+    // clear former jacobian elements
+    m_jac_trips.clear();
+    // reserve space for jacobian elements in triplet vector
+    if (m_jac_trips.capacity() < m_nv * m_nv)
+    {
+        m_jac_trips.reserve(m_nv * m_nv);
+    }
+    // Determine Species Derivatives
+    // get ROP derivatives
+    double scalingFactor = m_vol/accumulate(LHS + m_sidx, LHS + m_nv, 0.0);
+    Eigen::SparseMatrix<double> speciesDervs = scalingFactor * m_kin->netProductionRates_ddX();
+    // add to preconditioner
+    for (int k=0; k<speciesDervs.outerSize(); ++k) {
+        for (Eigen::SparseMatrix<double>::InnerIterator it(speciesDervs, k); it; ++it) {
+            m_jac_trips.push_back(Eigen::Triplet<double>(it.row() + m_sidx, it.col() + m_sidx, it.value()));
+        }
+    }
+    // Temperature Derivatives
+    if (m_energy) {
+        // getting perturbed state for finite difference
+        double deltaTemp = LHS[0] * DBL_EPSILON;
+        // finite difference temperature derivatives
+        vector_fp yNext (m_nv);
+        vector_fp ydotNext (m_nv);
+        vector_fp yCurrent (m_nv);
+        vector_fp ydotCurrent (m_nv);
+        // copy LHS to current and next
+        copy(LHS, LHS + m_nv, yCurrent.begin());
+        copy(LHS, LHS + m_nv, yNext.begin());
+        // perturb temperature
+        yNext[0] += deltaTemp;
+        // getting perturbed state
+        updateState(yNext.data());
+        eval(t, yNext.data(), ydotNext.data());
+        // reset and get original state
+        updateState(yCurrent.data());
+        eval(t, yCurrent.data(), ydotCurrent.data());
+        // d T_dot/dT
+        m_jac_trips.push_back(Eigen::Triplet<double>(0, 0, (ydotNext[0] - ydotCurrent[0]) / deltaTemp));
```

Comment:
You can simplify this (and all other places where adding elements to `m_jac_trips`) by using `emplace_back`, e.g.
```suggestion
        m_jac_trips.emplace_back(0, 0, (ydotNext[0] - ydotCurrent[0]) / deltaTemp);
```